[http_request] Unable to handle chunked responses

HLFCode opened the pull request:
# What does this implement/fix?

Fix for both Arduino and ESP-IDF platforms unable to handle chunked responses.
The ESP-IDF client intrinsically implements `Transfer-Encoding: chunked` it but wasn't enabled in ESPHome.
The Arduino client has no logic for decoding chunks.
This fix enables chunked responses in both platforms.

Tested with chunks of various lengths and with small  to large responses.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes [http_request.get broken(?) ](https://github.com/esphome/issues/issues/5949)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

bdraco commented on esphome/components/http_request/http_request.h:
```diff
@@ -234,19 +236,39 @@ template<typename... Ts> class HttpRequestSendAction : public Action<Ts...> {
       ExternalRAMAllocator<uint8_t> allocator(ExternalRAMAllocator<uint8_t>::ALLOW_FAILURE);
       uint8_t *buf = allocator.allocate(max_length);
       if (buf != nullptr) {
+        // // temporary initialiser for debugging (i.e. printing as a string)
+        // for (int i=0; i<max_length; i++) {
+        //   *(buf + i) = 0x0;
+        // }
```
Leftover debug code?

HLFCode replied:
Leftover as an aid for future developers - do you want me to remove it?

martinkirsche replied:
I removed that part in the name of @HLFCode  as commented out code will usually degrade over time and might never be useful again.

bdraco commented on esphome/components/http_request/http_request_arduino.cpp:
```diff
@@ -155,21 +170,117 @@ int HttpContainerArduino::read(uint8_t *buf, size_t max_len) {
+    int last_read_index = -1;  // value of read_index after the last read
+    while (read_len < bytes_to_read && read_len != last_read_index) {
+      last_read_index = read_len;
+      // limit this read count to the buffer size
+      // TODO: replace 512 with the h/w buffer size (whatever that is)
```
TODO?

HLFCode replied:
Yes, TODO
I put that comment in because I wanted to note that a future improvement would be to change the 512 with the actual size of the buffer.
Is it a problem?

martinkirsche replied:
@HLFCode why did you need to limit it? It looks like reading everything at once is fine when we got a none-chunked response, so why can't we read as much as we can for chunked responses?

bdraco commented:
We could write an integration test that spins up an aiohttp server and makes a request to it using this

HLFCode commented:
Hold off here please
I'm **so** confused with github my pull request is in a mess.
I've tried to get it up to date with the current dev branch but I think I've really messed it up.
It certainly doesn't work with the latest dev.

How can I delete everything and start again?

bdraco commented:
The PR currently has conflicts so it won't get a review until the CI is passing with conflicts resolved.

CloCkWeRX commented:
(Conflicts again)

CloCkWeRX commented:
Maybe https://github.com/esphome/esphome/pull/9353/files is also suitable

martinkirsche commented:
> Any clue how to take this forward @CloCkWeRX ? I see the pull request you linked is also > 200 comits behind.

I rebased this branch onto the latest `dev` branch in my fork at https://github.com/martinkirsche/esphome/tree/chunked_http_request

@HLFCode would you allow me to force push to your fork to update this PR (as far as I got it you have to allow me access via the [settings of your fork](https://github.com/HLFCode/esphome/settings/access))? Alternatively I can also create another PR with your changes.

martinkirsche commented on esphome/components/http_request/http_request.h:
```diff
@@ -253,16 +259,33 @@ template<typename... Ts> class HttpRequestSendAction : public Action<Ts...> {
+            // Read buffer too small
+            break;
+          }
+          int read = container->read(buf + read_index, max_length - read_index);
```
@HLFCode why did you not keep the `std::min<size_t>(max_length - read_index, 512)`?

I would guess this had been used for a reason before (probably to avoid letting the watchdog starve).

HLFCode replied:
> @HLFCode why did you not keep the `std::min<size_t>(max_length - read_index, 512)`?

It's been a really long time since I looked at this code so I'm not sure. If you see the need to change it please do.

>@HLFCode why did you need to limit it? It looks like reading everything at once is fine when we got a none-chunked response, so why can't we read as much as we can for chunked responses?

Again, long time ago but my instinct says we can't read everything at once if the sent data is larger than the hardware buffer (potential overflow)

If you'd prefer to copy/own this code and tweak it I'm happy with that - the important thing is to get it into the core code asap. I don't mind being dropped off the credit list.

I've not seen any comments about problems with the code so I'm assuming the code works as-is for small to large chunks, and small and large unchunked files - it does for me

I'm very grateful for your interest and help as my knowledge of GIT/Github is zip.

martinkirsche replied:
While rebasing your changes on the latest `dev` branch was quick and easy, I am not sure if I can find the time to fine-tune the implementation.

Sadly I also just discovered that with your changes a simple `http_request.get: url: "http://google.de"` stops working, so we can't merge the changes as they are.

martinkirsche replied:
I found some time to go over the changes and simplified them, because we don't actually need to calculate the amount of bytes to read (I added some details to the commit message).

 I also removed the Arduino based implementation as I can't test it anymore using my board after https://github.com/esphome/esphome/pull/12428 got merged (I can add the changes again if anyone volunteers to test and fix them).

I used this Yaml for testing:

```yaml
esphome:
  name: http_test

esp32:
  board: seeed_xiao_esp32c3
  framework:
    type: esp-idf

logger:
  level: VERY_VERBOSE

http_request:
  useragent: esphome
  timeout: 2s
  verify_ssl: false

.http_get: &http_get
  capture_response: true
  max_response_buffer_size: 512
  on_response: 
    then:
      - logger.log:
          format: "Response status: %d, Content-Length: %s (header) %d (stream) body: %s"
          args:
            - response->status_code
            - response->get_response_header("Content-Length").c_str()
            - response->content_length
            - body.c_str()

wifi:
  networks:
  - ssid: !secret wifi_ssid
    password: !secret wifi_password
  
  on_connect:
      # with redirect and a final response that is too large (even so the header says otherwise for some reason)
      - http_request.get: 
          <<: *http_get
          url: "https://google.de"
      
      # no redirect and the response fits inside the buffer    
      - http_request.get: 
          <<: *http_get
          url: "https://xkcd.com/3188/info.0.json"
      
      # chunked response (enabled by this fix)          
      - http_request.get: 
          <<: *http_get
          url: "http://widgets.vvo-online.de/abfahrtsmonitor/Abfahrten.do?hst=postplatz&ort=Dresden&lim=5"
```

This is the logging that the above test is producing for the latest `dev` commit (`c29aa61e`):

```
[00:48:49.881][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[00:48:49.883][D][http_request.idf:042]: Received response header, name: content-length, value: 219
[00:48:49.888][V][http_request.idf:178]: redirecting to url: https://www.google.de:443/
[00:48:51.597][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[00:48:51.598][D][http_request.idf:042]: Received response header, name: content-length, value: 220
[00:48:51.602][V][http_request.idf:178]: redirecting to url: https://www.google.com:443/
[00:48:53.322][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=ISO-8859-1
[00:48:53.326][D][http_request:038]: Header with name content-length found with value 219
[00:48:53.327][D][main:258]: Response status: 200, Content-Length: 219 (header) 0 (stream) body: 
[00:48:54.303][D][http_request.idf:042]: Received response header, name: content-length, value: 338
[00:48:54.305][D][http_request.idf:042]: Received response header, name: content-type, value: application/json
[00:48:54.310][D][http_request:038]: Header with name content-length found with value 338
[00:48:54.314][D][main:268]: Response status: 200, Content-Length: 338 (header) 338 (stream) body: {"month": "12", "num": 3188, "link": "", "year": "2025", "news": "", "safe_title": "Anyone Else Here", "transcript": "", "alt": "Anyone else watching this Youtube video in 1954? If so, my last trip definitely messed with the timeline.", "img": "https://imgs.xkcd.com/comics/anyone_else_here.png", "title": "Anyone Else Here", "day": "31"}
[00:48:54.431][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=utf-8
[00:48:54.433][W][http_request:028]: No header with name content-length found
[00:48:54.434][D][main:278]: Response status: 200, Content-Length:  (header) 0 (stream) body: 
````

The same code will return the following log with the fix of this PR (`0e305fbf`):

```
[00:46:08.885][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[00:46:08.887][D][http_request.idf:042]: Received response header, name: content-length, value: 219
[00:46:08.892][V][http_request.idf:178]: redirecting to url: https://www.google.de:443/
[00:46:10.603][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[00:46:10.605][D][http_request.idf:042]: Received response header, name: content-length, value: 220
[00:46:10.609][V][http_request.idf:178]: redirecting to url: https://www.google.com:443/
[00:46:12.315][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=ISO-8859-1
[00:46:12.323][D][http_request:038]: Header with name content-length found with value 219
[00:46:12.329][D][main:258]: Response status: 200, Content-Length: 219 (header) 512 (stream) body: <!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="de"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="a_4ZRAAUuKoN1RMtgDGFcQ">(function(){var _g={kEI:'RKpZaeGyFNa6i-gPyL2u4QQ',kEXPI:'0,203049,1101176,5,2935815,94918,344796,5532829,45,368
[00:46:13.315][D][http_request.idf:042]: Received response header, name: content-length, value: 338
[00:46:13.316][D][http_request.idf:042]: Received response header, name: content-type, value: application/json
[00:46:13.324][D][http_request:038]: Header with name content-length found with value 338
[00:46:13.327][D][main:268]: Response status: 200, Content-Length: 338 (header) 338 (stream) body: {"month": "12", "num": 3188, "link": "", "year": "2025", "news": "", "safe_title": "Anyone Else Here", "transcript": "", "alt": "Anyone else watching this Youtube video in 1954? If so, my last trip definitely messed with the timeline.", "img": "https://imgs.xkcd.com/comics/anyone_else_here.png", "title": "Anyone Else Here", "day": "31"}
[00:46:13.460][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=utf-8
[00:46:13.462][W][http_request:028]: No header with name content-length found
[00:46:13.465][D][main:278]: Response status: 200, Content-Length:  (header) 131 (stream) body: [["3","Wilder Mann",""],["68","Cossebaude","1"],["7","Weixdorf","8"],["68","Leubnitzer Höhe","14"],["68","Leubnitzer Höhe","15"]]
```

I hope that this is sufficient for testing such a change.

HLFCode committed:
[http_request] Add support for chunked HTTP responses

This change removes the unnecessary trimming of buffer length, which
enables reading responses without knowing the actual length, first
(chunked responses).

When providing a buffer to read from a stream it is only necessary to
specify the length of the buffer. The actual amount of available data to
read is usually known by the stream. It won't keep waiting for more data
when the stream had already been closed or all data were received.

tr1plus commented:
Hi @martinkirsche / @pvizeli 

I see this is technically part of the new release now? But I still think it's not working.

the api I want to call (zabbix monitoring api) does not return content lenght, so now I get errors again when doing API calls from esp home, resulting in failed json parse

Did this one really get fixed?

martinkirsche commented:
> Hi @martinkirsche / @pvizeli
> 
> I see this is technically part of the new release now? But I still think it's not working.

@tr1plus no, it was not yet fixed on the `dev` branch. There was some related change (causing a conflict) but as an empty content-length header still prevents the response body from getting loaded, chunked responses will still not work (see my example above).

> 
> the api I want to call (zabbix monitoring api) does not return content lenght, so now I get errors again when doing API calls from esp home, resulting in failed json parse

If you would provide me with an example request for that API (e.g. as curl command), I would add it to my test case Yaml.

CloCkWeRX commented on esphome/components/http_request/http_request.h:
```diff
@@ -265,6 +263,12 @@ template<typename... Ts> class HttpRequestSendAction : public Action<Ts...> {
       }
+
+      if (container->content_length == 0) {
+        // For chunked responses the content-length header is missing, so we update the content_length with the total
```
Could this more simply check `esp_http_client_is_chunked` earlier on and set a flag appropriately?

https://github.com/esphome/esphome/blob/0e305fbfaa24a807102aad905f0bdf45ffdd69c2/esphome/components/http_request/http_request_idf.cpp#L155 will set this explicitly to 0 when it is chunked *or the header is missing*.

While the behaviour we want is probably the same in both cases; the comment at the moment focuses only on the chunked variation so is a bit misleading.

martinkirsche replied:
> Could this more simply check esp_http_client_is_chunked earlier on and set a flag appropriately?

Checking `esp_http_client_is_chunked` would mean that we also have to store that state within `container`. But we already got the `content_length = 0` as an indicator that it is unknown.

Should I add and use that extra property anyhow?

> While the behaviour we want is probably the same in both cases; the comment at the moment focuses only on the chunked variation so is a bit misleading.

Is it fine to only adopt the comment like this:

```suggestion
        // If the content-length header is missing (e.g. for chunked responses), update the content_length with the total
```

martinkirsche replied:
Maybe it would even be better to not update the `content_length` property and keep it at `0` for responses that are missing the `content-length` header.

The body length is as well available within `body.size()` and updating  `content_length` might be confusing as one might think that the header was present.

martinkirsche replied:
I applied that change as my test case still works fine when keeping `content_length` at `0`:

```
[12:00:33.856][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[12:00:33.858][D][http_request.idf:042]: Received response header, name: content-length, value: 219
[12:00:33.862][V][http_request.idf:178]: redirecting to url: https://www.google.de:443/
[12:00:35.595][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=UTF-8
[12:00:35.597][D][http_request.idf:042]: Received response header, name: content-length, value: 220
[12:00:35.601][V][http_request.idf:178]: redirecting to url: https://www.google.com:443/
[12:00:37.326][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=ISO-8859-1
[12:00:37.332][D][http_request:038]: Header with name content-length found with value 219
[12:00:37.337][D][main:286]: Response status: 200, Content-Length: 219 (header) 0 (stream) body: <!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="de"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="EkgNCRTCQlZzPfimrtYFeg">(function(){var _g={kEI:'VUhaaYvBEeyD9u8PiJjlmAw',kEXPI:'0,203005,45,1101174,6,2935815,78813,16105,78219,123645,1
[12:00:37.985][D][http_request.idf:042]: Received response header, name: content-length, value: 338
[12:00:37.987][D][http_request.idf:042]: Received response header, name: content-type, value: application/json
[12:00:37.991][D][http_request:038]: Header with name content-length found with value 338
[12:00:37.995][D][main:296]: Response status: 200, Content-Length: 338 (header) 338 (stream) body: {"month": "12", "num": 3188, "link": "", "year": "2025", "news": "", "safe_title": "Anyone Else Here", "transcript": "", "alt": "Anyone else watching this Youtube video in 1954? If so, my last trip definitely messed with the timeline.", "img": "https://imgs.xkcd.com/comics/anyone_else_here.png", "title": "Anyone Else Here", "day": "31"}
[12:00:38.128][D][http_request.idf:042]: Received response header, name: content-type, value: text/html; charset=utf-8
[12:00:38.130][W][http_request:028]: No header with name content-length found
[12:00:38.132][D][main:306]: Response status: 200, Content-Length:  (header) 0 (stream) body: [["11","Zschertnitz",""],["9","Kaditz","1"],["68","Cossebaude","1"],["3","Wilder Mann","3"],["1","Prohlis","3"]]
[12:00:39.094][D][http_request.idf:042]: Received response header, name: content-type, value: application/json; charset=utf-8
[12:00:39.096][W][http_request:028]: No header with name content-length found
[12:00:39.100][D][main:316]: Response status: 200, Content-Length:  (header) 0 (stream) body: {"latitude":0.0,"longitude":0.0,"generationtime_ms":0.04458427429199219,"utc_offset_seconds":0,"timezone":"GMT","timezone_abbreviation":"GMT","elevation":0.0,"current_units":{"time":"iso8601","interval":"seconds","uv_index":"","shortwave_radiation":"W/m²"},"current":{"time":"2026-01-04T11:00","interval":900,"uv_index":7.70,"shortwave_radiation":818.0}}
```

The example from https://github.com/esphome/esphome/issues/12812 also outputs the size perfectly fine by using `body.size()`:

```
[12:01:29.580][D][http_request.idf:042]: Received response header, name: content-type, value: application/json; charset=utf-8
[12:01:29.586][D][http_request:038]: Header with name content-type found with value application/json; charset=utf-8
[12:01:29.589][D][main:070]: Response status: 200, Duration: 1043 ms, Content-Type: application/json; charset=utf-8
[12:01:29.589][D][main:071]: Body length: 355
[12:01:29.594][D][main:072]: Body: {"latitude":0.0,"longitude":0.0,"generationtime_ms":0.06902217864990234,"utc_offset_seconds":0,"timezone":"GMT","timezone_abbreviation":"GMT","elevation":0.0,"current_units":{"time":"iso8601","interval":"seconds","uv_index":"","shortwave_radiation":"W/m²"},"current":{"time":"2026-01-04T11:00","interval":900,"uv_index":7.70,"shortwave_radiation":818.0}}
```

martinkirsche commented:
@bdraco I executed the tests that you mentioned at https://github.com/esphome/esphome/pull/9353#issuecomment-3104970436 with the following modifications

- tests that use larger buffers got reduced to max. 32kb to avoid my tiny ESP32 c3 from running out of memory
- TC-004 needed a `max_response_buffer_size` to store the response data
- `http_request:watchdog_timeout` needs to be bigger than `http_request:timeout` to prevent the watchdog from timing out first
- `response->chunked` is not logged as it is not forwarded to the user (as the user already got the complete body, it should not be important anymore if the response was chunked or not)

Here are my modified test files:
- [http_test.yaml](https://github.com/user-attachments/files/24683594/http_test.yaml)
- [server.py](https://github.com/user-attachments/files/24683595/server.py)

And this is the log that my ESP32 c3 produces:
- [esp-log.txt](https://github.com/user-attachments/files/24683593/esp-log.txt)

I checked that I do not get any error or warning logs for any test case. Here are only the test case logs filtered from the log file linked above for reference:

```
[00:44:53.937][I][test_runner:337]: Starting all chunked transfer encoding tests...
[00:44:55.445][I][TC-001:043]: Response body length: 34
[00:44:55.446][I][TC-001:046]: Response content_length: 0
[00:44:55.446][I][TC-001:047]: Response bytes_read: 34
[00:44:55.446][I][TC-001:048]: Response status_code: 200
[00:44:55.447][I][TC-001:049]: Response duration_ms: 512
[00:44:55.448][I][TC-001:053]: Transfer-Encoding header: chunked
[00:44:55.449][I][TC-001:058]: Response body: Hello, this is a chunked response!
[00:44:57.975][I][TC-002:076]: Large file received, size: 32736 bytes
[00:44:57.976][I][TC-002:079]: SUCCESS: Received complete 32KB file
[00:45:03.448][I][TC-003:098]: Response body length: 46
[00:45:03.448][I][TC-003:100]: Status code: 200
[00:45:03.449][I][TC-003:101]: Duration: 457 ms
[00:45:03.450][I][TC-003:105]: Small chunks response: This is a test message with very small chunks!
[00:45:05.948][I][TC-004:123]: Variable chunks received, total size: 1665 bytes
[00:45:05.949][I][TC-004:125]: Expected: 1665 bytes (10+100+1000+50+500+5)
[00:45:05.950][I][TC-004:127]: SUCCESS: Received all variable chunks
[00:45:08.442][I][TC-005:148]: Response terminated correctly: First chunkSecond chunkFinal chunk
[00:46:48.697][I][TC-008:196]: Non-chunked response: This is a standard response with Content-Length header
[00:46:50.953][I][TC-009:216]: Response body: {"message": "Part 1"}{"data": [1, 2, 3]}{"status": "complete"}
[00:46:50.955][I][TC-009:217]: Content-Type: application/json
[00:46:50.957][I][TC-009:218]: X-Custom-Header: test-value
[00:46:50.958][I][TC-009:219]: Cache-Control: no-cache
[00:46:53.020][I][TC-010:235]: Buffer boundary test passed, size: 1000
[00:46:55.309][I][TC-011:249]: Performance test: 30000 bytes received
[00:46:55.309][I][TC-011:251]: Transfer time: 203 ms
[00:46:55.310][I][TC-011:254]: Transfer rate: 144.32 KB/s
[00:46:58.870][I][TC-012:271]: Mixed content received, size: 323 bytes
[00:46:58.871][D][TC-012:274]: Byte 0: 0x54
[00:46:58.871][D][TC-012:274]: Byte 1: 0x65
[00:46:58.872][D][TC-012:274]: Byte 2: 0x78
[00:46:58.872][D][TC-012:274]: Byte 3: 0x74
[00:46:58.873][D][TC-012:274]: Byte 4: 0x20
[00:46:58.873][D][TC-012:274]: Byte 5: 0x64
[00:46:58.873][D][TC-012:274]: Byte 6: 0x61
[00:46:58.874][D][TC-012:274]: Byte 7: 0x74
[00:46:58.874][D][TC-012:274]: Byte 8: 0x61
[00:46:58.875][D][TC-012:274]: Byte 9: 0x3A
[00:46:58.875][D][TC-012:274]: Byte 10: 0x20
[00:46:58.876][D][TC-012:274]: Byte 11: 0x48
[00:46:58.876][D][TC-012:274]: Byte 12: 0x65
[00:46:58.877][D][TC-012:274]: Byte 13: 0x6C
[00:46:58.877][D][TC-012:274]: Byte 14: 0x6C
[00:46:58.878][D][TC-012:274]: Byte 15: 0x6F
[00:46:58.878][D][TC-012:274]: Byte 16: 0x20
[00:46:58.879][D][TC-012:274]: Byte 17: 0x57
[00:46:58.879][D][TC-012:274]: Byte 18: 0x6F
[00:46:58.880][D][TC-012:274]: Byte 19: 0x72
[00:46:58.880][D][TC-012:274]: Byte 20: 0x6C
[00:46:58.880][D][TC-012:274]: Byte 21: 0x64
[00:46:58.881][D][TC-012:274]: Byte 22: 0x0A
[00:46:58.882][D][TC-012:274]: Byte 23: 0x00
[00:46:58.882][D][TC-012:274]: Byte 24: 0x01
[00:46:58.883][D][TC-012:274]: Byte 25: 0x02
[00:46:58.883][D][TC-012:274]: Byte 26: 0x03
[00:46:58.883][D][TC-012:274]: Byte 27: 0xFF
[00:46:58.884][D][TC-012:274]: Byte 28: 0xFE
[00:46:58.884][D][TC-012:274]: Byte 29: 0xFD
[00:46:58.885][D][TC-012:274]: Byte 30: 0x0A
[00:46:58.885][D][TC-012:274]: Byte 31: 0x4D
[00:47:01.039][I][TC-013:292]: Large single chunk received: 16384 bytes
[00:47:03.449][I][TC-014:308]: Response with extensions: Hello World!
[00:47:05.700][I][TC-015:327]: Response body: Data with trailing headers
[00:47:05.701][I][TC-015:328]: Trailer header: X-Checksum
[00:47:05.702][I][TC-015:329]: X-Checksum header: 
[00:47:07.706][I][test_runner:372]: All tests completed!
```

Can you try to reproduce this on your side? Maybe you even got a bigger ESP32 that can handle the large buffers.

bdraco committed:
Merge branch 'dev' into main

bdraco commented on esphome/components/http_request/http_request_idf.cpp:
```diff
@@ -213,18 +213,12 @@ int HttpContainerIDF::read(uint8_t *buf, size_t max_len) {
-
   this->feed_wdt();
-  int read_len = esp_http_client_read(this->client_, (char *) buf, bufsize);
+  int read_len = esp_http_client_read(this->client_, (char *) buf, max_len);
```
looks like caller will handle 0 ok

bdraco commented on esphome/components/http_request/http_request_idf.cpp:
```diff
@@ -213,18 +213,12 @@ int HttpContainerIDF::read(uint8_t *buf, size_t max_len) {
   const uint32_t start = millis();
   watchdog::WatchdogManager wdm(this->parent_->get_watchdog_timeout());
 
-  int bufsize = std::min(max_len, this->content_length - this->bytes_read_);
-
-  if (bufsize == 0) {
-    this->duration_ms += (millis() - start);
-    return 0;
-  }
-
```
esp_http_client_read handles this internally so should be fine to remove

bdraco replied:
https://github.com/espressif/esp-idf/blob/v5.5.2/components/esp_http_client/esp_http_client.c#L1325-L1332

bdraco reviewed (approved):
Thanks @HLFCode

tjhowse commented:
Awesome! Thanks @HLFCode, @bdraco and everyone else involved for helping this over the line. I look forward to being able to remove the hacks from my project and go back to mainline esphome.

HLFCode commented:
Thanks @martinkirsche @bdraco - a real success
Excuse my Github ignorance but can I safely delete this fork now?

bdraco commented:
> Thanks @martinkirsche @bdraco - a real success Excuse my Github ignorance but can I safely delete this fork now?

Yes
